Foundation/URLSession/BodySource.swift

```diff
@@ -57,6 +57,39 @@ internal enum _BodySourceDataChunk {
     case error
 }
 
+internal final class _BodyStreamSource {
+    let inputStream: InputStream
+    
+    init(inputStream: InputStream) {
+        self.inputStream = inputStream
+    }
+}
+
+extension _BodyStreamSource : _BodySource {
+    func getNextChunk(withLength length: Int) -> _BodySourceDataChunk {
+        guard inputStream.hasBytesAvailable else {
+            return .done
+        }
+        
+
+        let buffer = UnsafeMutableRawBufferPointer.allocate(count: length)
```

Comment:
When is it deallocated? 🤔

Is it possible to have `length > 0` and  `readBytes <= 0` ?

Reply:
@ogres thanks for you good catch! 👏

> When is it deallocated? 🤔

It looks like a memory leak, and we need to dealloc `buffer`. We can use `DispatchData(bytesNoCopy: UnsafeRawBufferPointer(buffer), deallocator: .free)` - it will free a memory and also do not creates a copy of bytes.

> Is it possible to have length > 0 and readBytes <= 0 ?

Yes, It can be, and unfortunately, we cannot be sure that data available during all time. Data and socket have different lifecycles so it’s ok that data is not available but the socket is alive and it wants a new chunk of data.

Reply:
@ogres I'll try it ASAP, I stuck with building Foundation for OSX.